README
= georuby-extras

* http://github.com/davetroy/georuby-extras

== DESCRIPTION:

GeoRuby-Extras Gem for Ruby (c) 2008 David Troy
dave@roundhousetech.com

GeoRuby-Extras offers some native C enhancements and extras to the popular GeoRuby gem.

GeoRuby offers OGC-compliant simple features for Ruby as well as the ability to
process SHP files and do other cool GIS hacks.  GeoRuby is often used in conjunction
with GIS extensions for popular databases including Postgres and MySQL.

== FEATURES/PROBLEMS:

* Provides native C implementation of Vincenty distance and direct algorithms.
* 25 times faster than GeoRuby 'ellipsoidal_distance' function.
* Adds the Vincenty 'direct' algorithm (point_at_bearing_and_distance).
* No known problems; won't work with JRuby or non-MRI Ruby implementations.

== SYNOPSIS:

require 'rubygems'
require 'georuby-extras'

include GeoRuby::SimpleFeatures
  
@home = Point.from_lon_lat(-76.511,39.024)
@nyc = Point.from_lon_lat(-74,40.6)

>> @home.ellipsoidal_distance(@nyc)
=> 277195.510448391

>> @home.point_at_bearing_and_distance(91.0, 130000.0).kml_representation
=> "<Point>\n<coordinates>-75.0105920700835,38.9939120193168</coordinates>\n</Point>\n"

== REQUIREMENTS:

* GCC and a Gnu-ish build environment (for native extensions)

== INSTALLATION

1) Enable gems from github, if you haven't already done so (rubygems >= 1.2):
    > sudo gem sources -a http://gems.github.com

2) Install gem
    > sudo gem install davetroy-georuby-extras

3) Profit!

== LICENSE:

(The MIT License)

Copyright (c) 2008 David Troy, Roundhouse Technologies LLC

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 0.5.0 / 2008-12-3

  * Released by David Troy (C) Under the MIT License


test/benchmark.rb
require './test_vicenty.rb'
require 'benchmark'

def test_multiple
  Benchmark.bmbm(30) do |bm|
    bm.report("   c distance") {30000.times { test_distance }}
    bm.report("ruby distance") {30000.times { test_orig_distance }}
  end
end

test/test_vicenty.rb
#!/usr/bin/env ruby
require 'rubygems'
require 'test/unit'
require 'geo_ruby'
require "#{File.dirname(__FILE__)}/../lib/georuby-extras"

class GeoHashNativeTest < Test::Unit::TestCase
  
  include GeoRuby::SimpleFeatures

  def setup
    @home = Point.from_lon_lat(-76.511,39.024)
    @nyc = Point.from_lon_lat(-73.5,40.2)
  end
  
  def test_distance
    assert_in_delta 289643.11, @home.ellipsoidal_distance(@nyc), 1
  end
  
  def test_orig_distance
    assert_in_delta 289643.11, @home.orig_ellipsoidal_distance(@nyc), 1
  end
     
  def test_bearing_from_point
    dest = @home.point_at_bearing_and_distance(91.0, 130000.0)
    assert_in_delta -75.0106, dest.x, 0.0001
    assert_in_delta 38.9939, dest.y, 0.0001
  end

end

test/test_vincenty.rb
#!/usr/bin/env ruby
require 'rubygems'
require 'test/unit'
require 'geo_ruby'
require "#{File.dirname(__FILE__)}/../lib/georuby-extras"

class GeoHashNativeTest < Test::Unit::TestCase
  
  include GeoRuby::SimpleFeatures

  def setup
    @home = Point.from_lon_lat(-76.511,39.024)
    @nyc = Point.from_lon_lat(-74,40.6)
  end
  
  def test_distance
    assert_in_delta 277195.5104, @home.ellipsoidal_distance(@nyc), 0.001
  end
  
  def test_orig_distance
    assert_in_delta 277195.5104, @home.orig_ellipsoidal_distance(@nyc), 0.001
  end

  def test_bearing_from_point
    dest = @home.point_at_bearing_and_distance(91.0, 130000.0)
    assert_in_delta -75.0106, dest.x, 0.0001
    assert_in_delta 38.9939, dest.y, 0.0001
  end
  
  # This test is basically identical to those in georuby; to confirm compatibility
  def test_point_distance
    point1 = Point.from_x_y(0,0)
    point2 = Point.from_x_y(3,4)
    assert_equal(5,point1.euclidian_distance(point2))
    
    assert_in_delta(554058.924,point1.ellipsoidal_distance(point2),0.001)
    assert_in_delta(554058.924,point1.orig_ellipsoidal_distance(point2),0.001)
         
    assert_in_delta(555811.68,point1.spherical_distance(point2),0.01)
  end
  
end

ext/vicenty.c
// geohash-native.c
// (c) 2008 David Troy
// dave@roundhousetech.com
//
// Based on code by Chris Veness
// http://www.movable-type.co.uk/scripts/latlong-vincenty.html
// http://www.movable-type.co.uk/scripts/latlong-vincenty-direct.html
// 
// Derived from algorithms by T. Vicenty, 1975
//
// (The MIT License)
// 
// Copyright (c) 2008 David Troy, Roundhouse Technologies LLC
// 
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// 'Software'), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
// 
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

#include "ruby.h"
#include <math.h>
#include <ctype.h>

#define DEG_TO_RAD	0.0174532925199433

static VALUE rb_mVicenty;

static VALUE distance(VALUE self, VALUE rb_lon1, VALUE rb_lat1, VALUE rb_lon2, VALUE rb_lat2, VALUE rb_a, VALUE rb_b) {

	Check_Type(rb_lon1, T_FLOAT);
	Check_Type(rb_lat1, T_FLOAT);
	Check_Type(rb_lon2, T_FLOAT);
	Check_Type(rb_lat2, T_FLOAT);
	Check_Type(rb_a, T_FLOAT);
	Check_Type(rb_b, T_FLOAT);

	double lon1 = RFLOAT(rb_lon1)->value;
	double lat1 = RFLOAT(rb_lat1)->value;
	double lon2 = RFLOAT(rb_lon2)->value;
	double lat2 = RFLOAT(rb_lat2)->value;
	double a = RFLOAT(rb_a)->value;
	double b = RFLOAT(rb_b)->value;

	double sinLambda, cosLambda, sinSigma, cosSigma, sigma, sinAlpha, cosSqAlpha, cos2SigmaM, C;
	
	double f = (a-b) / a;
  double L = (lon2-lon1) * DEG_TO_RAD;
  double U1 = atan((1-f) * tan(lat1 * DEG_TO_RAD));
  double U2 = atan((1-f) * tan(lat2 * DEG_TO_RAD));
	double sinU1 = sin(U1), cosU1 = cos(U1);
  double sinU2 = sin(U2), cosU2 = cos(U2);

  double lambda = L, lambdaP=2*3.14159, iterLimit = 20;
  while ((fabs(lambda-lambdaP) > 1e-12 && --iterLimit>0)) {
		sinLambda = sin(lambda); cosLambda = cos(lambda);
    sinSigma = sqrt((cosU2*sinLambda) * (cosU2*sinLambda) + (cosU1*sinU2-sinU1*cosU2*cosLambda) * (cosU1*sinU2-sinU1*cosU2*cosLambda));
    if (sinSigma==0) return rb_float_new(0);  // co-incident points
    cosSigma = sinU1*sinU2 + cosU1*cosU2*cosLambda;
    sigma = atan2(sinSigma, cosSigma);
    sinAlpha = cosU1 * cosU2 * sinLambda / sinSigma;
    cosSqAlpha = 1 - sinAlpha*sinAlpha;
    cos2SigmaM = cosSigma - 2*sinU1*sinU2/cosSqAlpha;
    if (isnan(cos2SigmaM)) cos2SigmaM = 0;  // equatorial line: cosSqAlpha=0 (§6)
    C = f/16*cosSqAlpha*(4+f*(4-3*cosSqAlpha));
    lambdaP = lambda;
    lambda = L + (1-C) * f * sinAlpha * (sigma + C*sinSigma*(cos2SigmaM+C*cosSigma*(-1+2*cos2SigmaM*cos2SigmaM)));
  }

	if (iterLimit==0) return Qnil;  // formula failed to converge

  double uSq = cosSqAlpha * (a*a - b*b) / (b*b);
  double A = 1 + uSq/16384.0*(4096+uSq*(-768+uSq*(320-175*uSq)));
  double B = uSq/1024 * (256+uSq*(-128+uSq*(74-47*uSq)));
  double deltaSigma = (B*sinSigma*(cos2SigmaM+B/4*(cosSigma*(-1+2*cos2SigmaM*cos2SigmaM)) -
    (B/6*cos2SigmaM*(-3+4*sinSigma*sinSigma)*(-3+4*cos2SigmaM*cos2SigmaM))));
	double s = b*A*(sigma-deltaSigma);
	
	return rb_float_new(s);
}

static VALUE point_from_lon_lat(VALUE self, VALUE rb_lon1, VALUE rb_lat1, VALUE rb_bearing, VALUE rb_distance, VALUE rb_a, VALUE rb_b) {
	Check_Type(rb_lon1, T_FLOAT);
	Check_Type(rb_lat1, T_FLOAT);
	Check_Type(rb_bearing, T_FLOAT);
	Check_Type(rb_distance, T_FLOAT);
	Check_Type(rb_a, T_FLOAT);
	Check_Type(rb_b, T_FLOAT);
	
	VALUE ret;

	double lon1 = RFLOAT(rb_lon1)->value;
	double lat1 = RFLOAT(rb_lat1)->value;
	double brng = RFLOAT(rb_bearing)->value;
	double s = RFLOAT(rb_distance)->value;
	double a = RFLOAT(rb_a)->value;
	double b = RFLOAT(rb_b)->value;
	
	double f = (a-b) / a;
  double alpha1 = brng * DEG_TO_RAD;
  double sinAlpha1 = sin(alpha1), cosAlpha1 = cos(alpha1);
  
  double tanU1 = (1-f) * tan(lat1 * DEG_TO_RAD);
  double cosU1 = 1 / sqrt((1 + tanU1*tanU1)), sinU1 = tanU1*cosU1;
  double sigma1 = atan2(tanU1, cosAlpha1);
  double sinAlpha = cosU1 * sinAlpha1;
  double cosSqAlpha = 1 - sinAlpha*sinAlpha;
  double uSq = cosSqAlpha * (a*a - b*b) / (b*b);
  double A = 1 + uSq/16384*(4096+uSq*(-768+uSq*(320-175*uSq)));
  double B = uSq/1024 * (256+uSq*(-128+uSq*(74-47*uSq)));
  
  double sigma = s / (b*A), sigmaP = 2*3.141592658;
	double cos2SigmaM, sinSigma, deltaSigma, cosSigma;
  while (fabs(sigma-sigmaP) > 1e-12) {
    cos2SigmaM = cos(2*sigma1 + sigma);
    sinSigma = sin(sigma), cosSigma = cos(sigma);
    deltaSigma = B*sinSigma*(cos2SigmaM+B/4*(cosSigma*(-1+2*cos2SigmaM*cos2SigmaM)-
      B/6*cos2SigmaM*(-3+4*sinSigma*sinSigma)*(-3+4*cos2SigmaM*cos2SigmaM)));
    sigmaP = sigma;
    sigma = s / (b*A) + deltaSigma;
  }

  double tmp = sinU1*sinSigma - cosU1*cosSigma*cosAlpha1;
  double lat2 = atan2(sinU1*cosSigma + cosU1*sinSigma*cosAlpha1, (1-f)*sqrt(sinAlpha*sinAlpha + tmp*tmp));
  double lambda = atan2(sinSigma*sinAlpha1, cosU1*cosSigma - sinU1*sinSigma*cosAlpha1);
  double C = f/16*cosSqAlpha*(4+f*(4-3*cosSqAlpha));
  double L = lambda - (1-C) * f * sinAlpha * (sigma + C*sinSigma*(cos2SigmaM+C*cosSigma*(-1+2*cos2SigmaM*cos2SigmaM)));

  double revAz = atan2(sinAlpha, -tmp);  // final bearing

	ret = rb_ary_new2(2); /* [lon, lat] */
	rb_ary_store(ret, 0, rb_float_new(lon1+L/DEG_TO_RAD));
	rb_ary_store(ret, 1, rb_float_new(lat2/DEG_TO_RAD));

	return ret;
}

void Init_vicenty()
{
	rb_mVicenty = rb_define_module("Vicenty");
	rb_define_module_function(rb_mVicenty, "distance", distance, 6);
	rb_define_module_function(rb_mVicenty, "point_from_lon_lat", point_from_lon_lat, 6);
}

// end


lib/rbvicenty.rb
module GeoRuby
  module SimpleFeatures
    #Represents a point. It is in 3D if the Z coordinate is not +nil+.
    class Point < Geometry
      def ellipsoidal_distance2(point, a = 6378137.0, b = 6356752.3142)
        deg_to_rad = 0.0174532925199433
        
        f = (a-b) / a
        l = (point.lon - lon) * deg_to_rad
        
        u1 = Math.atan((1-f) * Math.tan(lat * deg_to_rad ))
        u2 = Math.atan((1-f) * Math.tan(point.lat * deg_to_rad))
        sinU1 = Math.sin(u1)
        cosU1 = Math.cos(u1)
        sinU2 = Math.sin(u2)
        cosU2 = Math.cos(u2)
  
        puts "sinU1, cosU1 = #{sinU1}, #{cosU1}"
        puts "sinU2, cosU2 = #{sinU2}, #{cosU2}"
        puts "L = #{l}"
        lambda = l
        lambdaP = 2 * Math::PI
        iterLimit = 20
        
        while (lambda-lambdaP).abs > 1e-12 && --iterLimit>0
          sinLambda = Math.sin(lambda)
          cosLambda = Math.cos(lambda)
          puts "sinLambda, cosLambda = #{sinLambda}, #{cosLambda}"
          sinSigma = Math.sqrt((cosU2*sinLambda) * (cosU2*sinLambda) + (cosU1*sinU2-sinU1*cosU2*cosLambda) * (cosU1*sinU2-sinU1*cosU2*cosLambda))
          
          return 0 if sinSigma == 0 #coincident points
          
          cosSigma = sinU1*sinU2 + cosU1*cosU2*cosLambda
          sigma = Math.atan2(sinSigma, cosSigma)
          sinAlpha = cosU1 * cosU2 * sinLambda / sinSigma
          cosSqAlpha = 1 - sinAlpha*sinAlpha
          cos2SigmaM = cosSigma - 2*sinU1*sinU2/cosSqAlpha
          
          cos2SigmaM = 0 if (cos2SigmaM.nan?) #equatorial line: cosSqAlpha=0

          c = f/16*cosSqAlpha*(4+f*(4-3*cosSqAlpha))
          lambdaP = lambda
          lambda = l + (1-c) * f * sinAlpha * (sigma + c * sinSigma * (cos2SigmaM + c * cosSigma * (-1 + 2 * cos2SigmaM * cos2SigmaM)))
        end
        return NaN if iterLimit==0 #formula failed to converge

        puts "lambdaP, lambda = #{lambdaP}, #{lambda}"
        uSq = cosSqAlpha * (a*a - b*b) / (b*b)
        a_bis = 1 + uSq/16384*(4096+uSq*(-768+uSq*(320-175*uSq)))
        b_bis = uSq/1024 * (256+uSq*(-128+uSq*(74-47*uSq)))
        deltaSigma = b_bis * sinSigma*(cos2SigmaM + b_bis/4*(cosSigma*(-1+2*cos2SigmaM*cos2SigmaM)- b_bis/6*cos2SigmaM*(-3+4*sinSigma*sinSigma)*(-3+4*cos2SigmaM*cos2SigmaM)))
      
        b*a_bis*(sigma-deltaSigma)
      end
    end
  end
end

lib/georuby-extras.rb
require 'geo_ruby'
require "#{File.dirname(__FILE__)}/../ext/vincenty"

module GeoRuby
  module SimpleFeatures
    class Point < Geometry
    
      alias_method :orig_ellipsoidal_distance, :ellipsoidal_distance

      # Implementation of the Vincenty 'distance' formula to find the ellipsoidal distance between points
      # As per implementation by Chris Veness (http://www.movable-type.co.uk/scripts/latlong-vincenty.html)
      # Default a/b values are for the WGS84 ellipsoid - other values may be specified
      def ellipsoidal_distance(point, a = 6378137.0, b = 6356752.3142)
        Vincenty.distance(x.to_f, y.to_f, point.x.to_f, point.y.to_f, a.to_f, b.to_f)
      end

      # Implementation of the Vincenty 'direct' formula to find a point based on a bearing and distance
      # As per implementation by Chris Veness (http://www.movable-type.co.uk/scripts/latlong-vincenty-direct.html)
      # Default a/b values are for the WGS84 ellipsoid - other values may be specified
      def point_at_bearing_and_distance(bearing=0.0, distance=0.0, a = 6378137.0, b = 6356752.3142)
        Point.from_coordinates(Vincenty.point_from_lon_lat(lon.to_f,lat.to_f,bearing.to_f,distance.to_f,a.to_f,b.to_f))
      end
      
    end
    
  end
end

ext/vincenty.c
// geohash-native.c
// (c) 2008 David Troy
// dave@roundhousetech.com
//
// Based on Javascript code (c) by Chris Veness
// http://www.movable-type.co.uk/scripts/latlong-vincenty.html
// http://www.movable-type.co.uk/scripts/latlong-vincenty-direct.html
// 
// Derived from algorithms by T. Vincenty, 1975
//
// (The MIT License)
// 
// Copyright (c) 2008 David Troy, Roundhouse Technologies LLC
// 
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// 'Software'), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
// 
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

#include "ruby.h"
#include <math.h>
#include <ctype.h>

#define DEG_TO_RAD	0.0174532925199433
#define PI					3.14159265358979

static VALUE rb_mVincenty;

static VALUE distance(VALUE self, VALUE rb_lon1, VALUE rb_lat1, VALUE rb_lon2, VALUE rb_lat2, VALUE rb_a, VALUE rb_b) {

	Check_Type(rb_lon1, T_FLOAT);
	Check_Type(rb_lat1, T_FLOAT);
	Check_Type(rb_lon2, T_FLOAT);
	Check_Type(rb_lat2, T_FLOAT);
	Check_Type(rb_a, T_FLOAT);
	Check_Type(rb_b, T_FLOAT);

	double lon1 = RFLOAT(rb_lon1)->value;
	double lat1 = RFLOAT(rb_lat1)->value;
	double lon2 = RFLOAT(rb_lon2)->value;
	double lat2 = RFLOAT(rb_lat2)->value;
	double a = RFLOAT(rb_a)->value;
	double b = RFLOAT(rb_b)->value;

	double sinLambda, cosLambda, sinSigma, cosSigma, sigma, sinAlpha, cosSqAlpha, cos2SigmaM, C;
	
	double f = (a-b) / a;
  double L = (lon2-lon1) * DEG_TO_RAD;
  double U1 = atan((1-f) * tan(lat1 * DEG_TO_RAD));
  double U2 = atan((1-f) * tan(lat2 * DEG_TO_RAD));
	double sinU1 = sin(U1), cosU1 = cos(U1);
  double sinU2 = sin(U2), cosU2 = cos(U2);

	int iterLimit = 20;
  double lambda = L, lambdaP=2*PI;
  while ((fabs(lambda-lambdaP) > 1e-12 && --iterLimit>0)) {
		sinLambda = sin(lambda); cosLambda = cos(lambda);
    sinSigma = sqrt((cosU2*sinLambda) * (cosU2*sinLambda) + (cosU1*sinU2-sinU1*cosU2*cosLambda) * (cosU1*sinU2-sinU1*cosU2*cosLambda));
    if (sinSigma==0) return rb_float_new(0);  // co-incident points
    cosSigma = sinU1*sinU2 + cosU1*cosU2*cosLambda;
    sigma = atan2(sinSigma, cosSigma);
    sinAlpha = cosU1 * cosU2 * sinLambda / sinSigma;
    cosSqAlpha = 1 - sinAlpha*sinAlpha;
    cos2SigmaM = cosSigma - 2*sinU1*sinU2/cosSqAlpha;
    if (isnan(cos2SigmaM)) cos2SigmaM = 0;  // equatorial line: cosSqAlpha=0 (§6)
    C = f/16*cosSqAlpha*(4+f*(4-3*cosSqAlpha));
    lambdaP = lambda;
    lambda = L + (1-C) * f * sinAlpha * (sigma + C*sinSigma*(cos2SigmaM+C*cosSigma*(-1+2*cos2SigmaM*cos2SigmaM)));
  }

	if (iterLimit==0) return Qnil;  // formula failed to converge

  double uSq = cosSqAlpha * (a*a - b*b) / (b*b);
  double A = 1 + uSq/16384.0*(4096+uSq*(-768+uSq*(320-175*uSq)));
  double B = uSq/1024 * (256+uSq*(-128+uSq*(74-47*uSq)));
  double deltaSigma = B*sinSigma*(cos2SigmaM+B/4*(cosSigma*(-1+2*cos2SigmaM*cos2SigmaM) -
    B/6*cos2SigmaM*(-3+4*sinSigma*sinSigma)*(-3+4*cos2SigmaM*cos2SigmaM)));

	double s = b*A*(sigma-deltaSigma);
	
	return rb_float_new(s);
}

static VALUE point_from_lon_lat(VALUE self, VALUE rb_lon1, VALUE rb_lat1, VALUE rb_bearing, VALUE rb_distance, VALUE rb_a, VALUE rb_b) {
	Check_Type(rb_lon1, T_FLOAT);
	Check_Type(rb_lat1, T_FLOAT);
	Check_Type(rb_bearing, T_FLOAT);
	Check_Type(rb_distance, T_FLOAT);
	Check_Type(rb_a, T_FLOAT);
	Check_Type(rb_b, T_FLOAT);
	
	VALUE ret;

	double lon1 = RFLOAT(rb_lon1)->value;
	double lat1 = RFLOAT(rb_lat1)->value;
	double brng = RFLOAT(rb_bearing)->value;
	double s = RFLOAT(rb_distance)->value;
	double a = RFLOAT(rb_a)->value;
	double b = RFLOAT(rb_b)->value;
	
	double f = (a-b) / a;
  double alpha1 = brng * DEG_TO_RAD;
  double sinAlpha1 = sin(alpha1), cosAlpha1 = cos(alpha1);
  
  double tanU1 = (1-f) * tan(lat1 * DEG_TO_RAD);
  double cosU1 = 1 / sqrt((1 + tanU1*tanU1)), sinU1 = tanU1*cosU1;
  double sigma1 = atan2(tanU1, cosAlpha1);
  double sinAlpha = cosU1 * sinAlpha1;
  double cosSqAlpha = 1 - sinAlpha*sinAlpha;
  double uSq = cosSqAlpha * (a*a - b*b) / (b*b);
  double A = 1 + uSq/16384*(4096+uSq*(-768+uSq*(320-175*uSq)));
  double B = uSq/1024 * (256+uSq*(-128+uSq*(74-47*uSq)));
  
  double sigma = s / (b*A), sigmaP = 2*PI;
	double cos2SigmaM, sinSigma, deltaSigma, cosSigma;
  while (fabs(sigma-sigmaP) > 1e-12) {
    cos2SigmaM = cos(2*sigma1 + sigma);
    sinSigma = sin(sigma), cosSigma = cos(sigma);
    deltaSigma = B*sinSigma*(cos2SigmaM+B/4*(cosSigma*(-1+2*cos2SigmaM*cos2SigmaM)-
      B/6*cos2SigmaM*(-3+4*sinSigma*sinSigma)*(-3+4*cos2SigmaM*cos2SigmaM)));
    sigmaP = sigma;
    sigma = s / (b*A) + deltaSigma;
  }

  double tmp = sinU1*sinSigma - cosU1*cosSigma*cosAlpha1;
  double lat2 = atan2(sinU1*cosSigma + cosU1*sinSigma*cosAlpha1, (1-f)*sqrt(sinAlpha*sinAlpha + tmp*tmp));
  double lambda = atan2(sinSigma*sinAlpha1, cosU1*cosSigma - sinU1*sinSigma*cosAlpha1);
  double C = f/16*cosSqAlpha*(4+f*(4-3*cosSqAlpha));
  double L = lambda - (1-C) * f * sinAlpha * (sigma + C*sinSigma*(cos2SigmaM+C*cosSigma*(-1+2*cos2SigmaM*cos2SigmaM)));

  double revAz = atan2(sinAlpha, -tmp);  // final bearing

	ret = rb_ary_new2(2); /* [lon, lat] */
	rb_ary_store(ret, 0, rb_float_new(lon1+L/DEG_TO_RAD));
	rb_ary_store(ret, 1, rb_float_new(lat2/DEG_TO_RAD));

	return ret;
}

void Init_vincenty()
{
	rb_mVincenty = rb_define_module("Vincenty");
	rb_define_module_function(rb_mVincenty, "distance", distance, 6);
	rb_define_module_function(rb_mVincenty, "point_from_lon_lat", point_from_lon_lat, 6);
}

// end
